### PR TITLE
feat(lint): add pre-commit linting setup

### PR DESCRIPTION
**Pull request by rriski**
Best reviewed commit by commit.

* run linting with [pre-commit](https://pre-commit.com/)
* add markdownlint check
* add codespell check
* add yamllint check
* add yamlfix check
* add hadolint check

Most of the diff is autofixed.

### CODE REVIEW

**Review by Serpentiel — CHANGES_REQUESTED** [review]
can we please have `markdownlint` added to the CI pipeline?

**Comment by Serpentiel** [comment]
also, please rename it to `docs(readme): run markdownlint` or something like that

File: .github/dependabot.yml [hunk 1]
```diff
@@ -1,4 +1,6 @@
+---
```
Comment (Serpentiel) [inline .github/dependabot.yml:1]: this isn't how YAML should start

**Review by Serpentiel — CHANGES_REQUESTED** [review]
WDYT about reusing our existing linting solutions, such as [this](https://github.com/aiven/go-api-schemas/blob/main/.trunk/trunk.yaml) one, which is then used [like that](https://github.com/aiven/go-api-schemas/blob/main/.github/workflows/lint.yml#L19-L29)?

File: .github/dependabot.yml (hunk 1 above)
Comment (rriski) [reply .github/dependabot.yml:1]: In fact it is, see section 2.2 of the YAML specification: https://yaml.org/spec/1.2.2/

Comment (Serpentiel) [reply .github/dependabot.yml:1]: It's only mandatory to have them if you use directives in the YAML document. Directives are optional for usage, please see [section 3.2.3.4](https://yaml.org/spec/1.2.2/#3234-directives) of the YAML specification. Please see examples in section [6.8.1](https://yaml.org/spec/1.2.2/#681-yaml-directives) to have a better understanding.

Comment (rriski) [reply .github/dependabot.yml:1]: Indeed its not mandatory and we can choose to leave it out. But stating "this isn't how YAML should start" isn't accurate.

**Comment by rriski** [comment]
> WDYT about reusing our existing linting solutions, such as [this](https://github.com/aiven/go-api-schemas/blob/main/.trunk/trunk.yaml) one, which is then used [like that](https://github.com/aiven/go-api-schemas/blob/main/.github/workflows/lint.yml#L19-L29)?

I don't mink trunk vs pre-commit. It looks like trunk doesn't support autofixable yaml linter which might be annoying. Trunk i a for-profit company whereas pre-commit is more open. It also looks like trunk is missing some of the nice [trailing whitespace, end-of-file, and shebang checks](https://github.com/aiven/aiven-operator/pull/573/commits/bf8f227e6ff9c90667840e65bbba68b3d46e0764#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9R6-R12) that come with pre-commit, didn't find any reference to something like that in [their docs](https://docs.trunk.io/check).Pre-commit also seems to be more widely used within Aiven.

I can create an alternate version of this PR that utilizes trunk.

**Comment by Serpentiel** [comment]
> looks like trunk doesn't support autofixable yaml linter which might be annoying

there's a YAML linter and it's autofixable, you can see it in the example that I linked to you, it's called `yamllint` (https://github.com/trunk-io/plugins/tree/main/linters/yamllint)

> Trunk i a for-profit company

it's free for Open Source usage :)

> [trailing whitespace, end-of-file, and shebang checks](https://github.com/aiven/aiven-operator/pull/573/commits/bf8f227e6ff9c90667840e65bbba68b3d46e0764#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9R6-R12)

trailing whitespace checks are included in `golangci-lint`; `end-of-file` and `shebang` checks aren't needed—it's very hard to justify need of an empty file, because those aren't generally needed/used, and `bash` files aren't used either because they're not cross-platform, while Go is

> Pre-commit also seems to be more widely used within Aiven.

in our Go projects it's not really used, and the only such solution that we ever used is Trunk

> I can create an alternate version of this PR that utilizes trunk.

sounds good; you can as well adjust this PR if you feel like it

**Comment by rriski** [comment]
> > looks like trunk doesn't support autofixable yaml linter which might be annoying
> 
> there's a YAML linter and it's autofixable, you can see it in the example that I linked to you, it's called `yamllint` (https://github.com/trunk-io/plugins/tree/main/linters/yamllint)
> 
> > Trunk i a for-profit company
> 
> it's free for Open Source usage :)
> 
> > [trailing whitespace, end-of-file, and shebang checks](https://github.com/aiven/aiven-operator/pull/573/commits/bf8f227e6ff9c90667840e65bbba68b3d46e0764#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9R6-R12)
> 
> trailing whitespace checks are included in `golangci-lint`; `end-of-file` and `shebang` checks aren't needed—it's very hard to justify need of an empty file, because those aren't generally needed/used, and `bash` files aren't used either because they're not cross-platform, while Go is
> 
> > Pre-commit also seems to be more widely used within Aiven.
> 
> in our Go projects it's not really used, and the only such solution that we ever used is Trunk
> 
> > I can create an alternate version of this PR that utilizes trunk.
> 
> sounds good; you can as well adjust this PR if you feel like it

Yep, makes sense. Alternate Trunk based approach can be found here: https://github.com/aiven/aiven-operator/pull/578.

**Comment by rriski** [comment]
Closed in favor of https://github.com/aiven/aiven-operator/pull/578

**Pull request closed without merging**